system/src/ble_listening_mode_handler.cpp

```diff
@@ -41,110 +42,99 @@ BleListeningModeHandler::BleListeningModeHandler()
           prePpcp_(),
           preAdvertising_(false),
           preConnected_(false),
-          preAutoAdv_(BLE_AUTO_ADV_ALWAYS) {
+          preAutoAdv_(BLE_AUTO_ADV_ALWAYS),
+          userAdv_(false),
+          ctrlReqAdvData_(nullptr),
+          ctrlReqAdvDataLen_(0),
+          ctrlReqSrData_(nullptr),
+          ctrlReqSrDataLen_(0) {
 }
 
 BleListeningModeHandler::~BleListeningModeHandler() {
 }
 
-int BleListeningModeHandler::enter() {
-    // TODO: Would it be better to call the method of BLE Control Request Channel to isolate the logic?
-    // Cache the application specific advertising data, advertising parameters and connection parameters.
+int BleListeningModeHandler::constructControlRequestAdvData() {
+    ctrlReqAdvData_.reset(new(std::nothrow) uint8_t[BLE_MAX_ADV_DATA_LEN]);
```

Comment:
Let's use the following pattern:

```cpp
{
    auto ctrlReqAdvData = std::make_unique<uint8_t[]>(BLE_MAX_ADV_DATA_LEN);
    CHECK_TRUE(ctrlReqAdvData);
    // ...
    auto ctrlReqSrData = std::make_unique<uint8_t[]>(BLE_MAX_ADV_DATA_LEN);
    CHECK_TRUE(ctrlReqSrData);
    // ...
    // A bunch of other stuff that may return from the function prematurely e.g. due to an error
    // ...
    ctrlReqAdvData_.swap(ctrlReqAdvData);
    ctrlReqSrData_.swap(ctrlReqSrData);
    return SYSTEM_ERROR_NONE;
}
```

Same thing in `cacheUserConfigurations()`.

Reply:
If we're going to use `Vector` instead, this should be different.

Reply:
Same pattern applies to the Vector. The following can also be used:
```cpp
ctrlReqAdvData_ = std::move(ctrlReqAdvData);
```

What I'm trying to suggest here is to fill in the member variables only at the end of the function.

Reply:
I see. In that case memory allocated in this function will be freed automatically if error occurs before copying the constructed data or cached data and the vector will empty then. 👍